status/k8s.go

```diff
@@ -477,12 +481,68 @@ func (k *K8sStatusCollector) status(ctx context.Context) *Status {
 		},
 	}
 
+	// for the sake of sanity, don't get pod logs more than once
+	var agentLogsOnce = sync.Once{}
 	err := k.podStatus(ctx, status, defaults.AgentDaemonSetName, "k8s-app=cilium", func(ctx context.Context, status *Status, name string, pod *corev1.Pod) {
 		if pod.Status.Phase == corev1.PodRunning {
+			// extract container status
+			var containerStatus *corev1.ContainerStatus
+			for i, cStatus := range pod.Status.ContainerStatuses {
+				if cStatus.Name == defaults.AgentContainerName {
+					containerStatus = &pod.Status.ContainerStatuses[i]
+					break
+				}
+			}
 			tasks = append(tasks, statusTask{
 				name: pod.Name,
-				task: func(_ context.Context) error {
-					s, err := k.client.CiliumStatus(ctx, k.params.Namespace, pod.Name)
+				task: func(ctx context.Context) error {
+					var s *models.StatusResponse
+					var err error
+
+					if containerStatus != nil && containerStatus.State.Running != nil {
+						// if container is running, execute "cilium status" in the container and parse the result
+
+						s, err = k.client.CiliumStatus(ctx, k.params.Namespace, pod.Name)
+
+					} else {
+						// otherwise, generate a useful status message
+						desc := "is not running"
+						lastLog := fmt.Sprintf("try 'kubectl -n %s logs -c %s %s'",
+							pod.Namespace, defaults.AgentContainerName, pod.Name)
+
+						// determine CrashLoopBackOff status and get last log line, if available.
+						if containerStatus != nil {
+							if containerStatus.State.Waiting != nil && containerStatus.State.Waiting.Reason == "CrashLoopBackOff" {
+								desc = "is in CrashLoopBackOff"
+							}
+							if containerStatus.LastTerminationState.Terminated != nil {
+								terminated := containerStatus.LastTerminationState.Terminated
+								desc = fmt.Sprintf("%s, exited with code %d", desc, terminated.ExitCode)
+
+								// capture final log line, maybe it's useful
+								// either from container message or a separate logs request
+								dyingGasp := ""
+								if terminated.Message != "" {
+									dyingGasp = strings.TrimSpace(terminated.Message)
+								} else {
+									agentLogsOnce.Do(func() { // in a sync.Once so we don't waste time retrieving lots of logs
+										logs, err := k.client.CiliumLogs(ctx, pod.Namespace, pod.Name, terminated.FinishedAt.Time.Add(-2*time.Minute), nil)
+										if err == nil && logs != "" {
+											dyingGasp = strings.TrimSpace(logs)
+										}
```

Comment:
If we fail to find logs the only time we enter this `func`, then we won't try another Pod correct? Not critical but not ideal either.

Reply:
Once the TerminationMessagePolicy PR goes in (https://github.com/cilium/cilium/pull/21012), this code path will be very unlikely indeed. So I doubt it's a big deal.